### PR TITLE
OF sequence number bug

### PR DESCRIPTION
**Pull request by szalpal**
Signed-off-by: Michał Szołucha <mszolucha@nvidia.com>

### CODE REVIEW

**Comment by szalpal** [comment]
!build

**Comment by kychennv** [comment]
is this for bug https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=200520245&cmtNo=?

**Comment by szalpal** [comment]
> is this for bug https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=200520245&cmtNo=?

Yes

**Comment by szalpal** [comment]
> Do we have any test for that?

@kychennv wrote it for the bug, will it be merged?

**Comment by JanuszL** [comment]
@szalpal - so please make sure that test is ready to be merged just after this PR.

File: dali/pipeline/operators/optical_flow/optical_flow.cc [hunk 1]
```diff
@@ -72,7 +72,7 @@ void OpticalFlow<GPUBackend>::RunImpl(Workspace<GPUBackend> *ws, const int) {
     std::vector<Dims> new_sizes;
     auto out_shape = optical_flow_->GetOutputShape();
     for (int i = 0; i < nsequences_; i++) {
-      Dims shape = {sequence_sizes_[i]};
+      Dims shape = {sequence_sizes_[i] - 1};
```
Comment (mzient) [inline dali/pipeline/operators/optical_flow/optical_flow.cc]:
```suggestion
      Dims shape = {sequence_sizes_[i] > 0 ? sequence_sizes_[i] - 1 : 0};
```

File: dali/pipeline/operators/optical_flow/optical_flow.cc [hunk 2]
```diff
@@ -114,7 +114,7 @@ void OpticalFlow<GPUBackend>::RunImpl(Workspace<GPUBackend> *ws, const int) {
...
```
Comment (mzient) [inline dali/pipeline/operators/optical_flow/optical_flow.cc]:
```suggestion
      Dims shape = {std::max(sequence_sizes_[i] - 1, 0)};
```

File: dali/pipeline/operators/optical_flow/optical_flow.cc (hunk 1 above)
Comment (klecki) [reply dali/pipeline/operators/optical_flow/optical_flow.cc]: Btw, shouldn't we enforce at least 2 elements in sequence?

Comment (mzient) [reply dali/pipeline/operators/optical_flow/optical_flow.cc]: ...and then a user comes with a broken dataset (vide SSD with no bboxes) and
(ノ°Д°）ノ︵ ┻━┻

**Comment by kychennv** [comment]
> > Do we have any test for that?
> 
> @kychennv wrote it for the bug, will it be merged?

yes,  but there is another bug exists: : http://nvbugs/200520501/

**Comment by szalpal** [comment]
> This would probably crash very badly for empty sequences.

There are 2 solutions:
1. Force sequences with at least 2 elements
2. Allow them and silently do workaround inside

I decided to go with the former. Calculating OF for sequence that's smaller than 2 elems is not really the use case and it's easier to spot than having bunch of empty OF results and thus NN not training properly

**Comment by szalpal** [comment]
!build

File: dali/pipeline/operators/optical_flow/optical_flow.h [hunk 3]
```diff
@@ -124,7 +124,14 @@ class OpticalFlow : public Operator<Backend> {
             return false;
         }
         return true;
-    }(), "Width, height and depth must be equal for all sequences");
+    }(), "Width, height and depth must be equal for all sequences.");
+
+    DALI_ENFORCE([&]() -> bool {
```
Comment (mzient) [inline dali/pipeline/operators/optical_flow/optical_flow.h]: Why not:
```
for (auto sz : sequence_sizes) {
  DALI_ENFORCE(sz >= 2, "Optical flow requires at least two frames");
}
```

**Comment by szalpal** [comment]
Confirmed, that it solves the error in (ref. #925 )

File: dali/pipeline/operators/optical_flow/optical_flow.h [hunk 4]
```diff
@@ -118,13 +118,14 @@ class OpticalFlow : public Operator<Backend> {
...
-    DALI_ENFORCE([&]() -> bool {
-        for (const auto &seq : shape) {
-          if (seq[1] != frames_height_ || seq[2] != frames_width_ || seq[3] != depth_)
-            return false;
-        }
-        return true;
-    }(), "Width, height and depth must be equal for all sequences");
+    for (const auto &seq : shape) {
+      DALI_ENFORCE(seq[1] != frames_height_ || seq[2] != frames_width_ || seq[3] != depth_,
+              "Width, height and depth must be equal for all sequences.");
```
Comment (szalpal) [inline dali/pipeline/operators/optical_flow/optical_flow.h]:
```suggestion
              "Width, height and depth for Optical Flow calculation must be equal for all sequences.");
```

File: dali/pipeline/operators/optical_flow/optical_flow.h [hunk 5]
```diff
@@ -118,13 +118,14 @@ class OpticalFlow : public Operator<Backend> {
...
+    }
+
+    for (auto sz : sequence_sizes_) {
+      DALI_ENFORCE(sz < 2, "Number of frames in every sequence must be greater or equal to 2.");
```
Comment (szalpal) [inline dali/pipeline/operators/optical_flow/optical_flow.h]:
```suggestion
      DALI_ENFORCE(sz < 2, "Optical Flow operator error: Number of frames in every sequence must be greater or equal to 2.");
```

File: dali/pipeline/operators/optical_flow/optical_flow.h [hunk 6]
```diff
@@ -118,13 +118,15 @@ class OpticalFlow : public Operator<Backend> {
...
-        return true;
-    }(), "Width, height and depth must be equal for all sequences");
+    for (const auto &seq : shape) {
+      DALI_ENFORCE(seq[1] != frames_height_ || seq[2] != frames_width_ || seq[3] != depth_,
+              "Width, height and depth for Optical Flow calculation must be equal for all sequences.");
+    }
+
+    for (auto sz : sequence_sizes_) {
+      DALI_ENFORCE(sz < 2, 
+        "Optical Flow operator error: Number of frames in every sequence must be greater or equal to 2.");
```
Comment (mzient) [inline dali/pipeline/operators/optical_flow/optical_flow.h]:
```suggestion
        ""At least two frames are necessary to calculate optical flow!");
```
It's shorter and explains the problem better.

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
Only lint failed. Fixed lint and merging.

**Pull request merged**